README.md
# Slack Driver

## About
This driver was created for the Mozilla IAM Project to satisfy an OKR around Slack user session expiration.

## Behavior

1. Spin up on cron/event trigger.
2. Scan the dynamodb table of all profiles.
3. Build a group data structure from all profiles.
4. Query the Slack API for all users profiles.
5. Fetch `apps.yml` access control file.
6. Disable any user without access to Slack through a Slack API call.
7. Enable any previously-disabled user that is still present in Slack database.


NOTE: If you have Slack owners (not admins), these cannot be deactivated. It is recommended to create service-accounts
for Slack owners, instead of using a normal / every-day user account. This is safer and cleaner, and you also will not
run into this issue that way since the owner accounts being service-accounts, will not need to be deactivated.

## Process Diagram
!['docs/img/Slack-Integration.png'](docs/img/Slack-Integration.png)

## Deployment

### Insert credstash api key

You only need to do this once.

```
credstash -r us-west-2 put -a slack-driver.token @slack-driver-api-key.txt app=slack-driver
```

To obtain the token, see <https://api.slack.com/scim> - TLDR:

1. Create an app at <https://api.slack.com/apps/new>
2. Click "set permissions" and add scope `admin`
3. Click "install app to workspace" and authorize at the prompt
4. Copy the "oauth access token" and keep it safe. That is your token for this program.


Note that you may restrict which IP ranges can call the API with this token in Slack's app settings as well.

### Deploy, test, etc

1. `cd slack_driver`
2. `make` for a list of targets, ex:

- `make python-venv` if you don't have your own virtual environment scripts

- `make tests` runs all tests
- `make deploy` deploys the code in the dev environment
- `make remove-deploy` deletes the dev deployment
- `make STAGE=prod deploy` deploys the code in the prod environment
- `make logs` just watch cloudwatch logs


CHANGELOG.md
# ChangeLog 
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-07-5

## Added
- Initial release


slack_driver/tests/test_slack_driver.py
#!/usr/bin/python

import driver
import json
import logging
import requests
import slack
import unittest

from unittest.mock import patch


logging.basicConfig(level=logging.DEBUG, format='%(asctime)s:%(levelname)s:%(name)s:%(message)s')
logger = logging.getLogger(__name__)


class SlackDriverTest(unittest.TestCase):
    def setUp(self):
        pass

    @patch.object(requests, 'get')
    def test_get_access_rules(self, mock_request):
        class fake_request(object):
            fake_json = {'apps': {'Slack': {'authorized_groups': ['team_moco']}}}
            text = json.dumps(fake_json)

            def ok(self):
                return True

        mock_request.return_value = fake_request
        driver.logger = logger
        ret = driver.get_access_rules('https://cdn.sso.mozilla.com/apps.yml')
        assert(ret['Slack']['authorized_groups'][0] == 'team_moco')

    @patch.object(driver, 'get_secret')
    @patch.object(slack.SlackAPI, 'get_users')
    @patch.object(slack.SlackAPI, 'deactivate_user')
    @patch.object(slack.SlackAPI, 'activate_user')
    def test_verify_slack_users(self, mock_slack_activate_user, mock_slack_deactivate_user, mock_slack_get_users,
                                mock_get_secret):
        fake_users = [{'active': False,
                       'emails': [{'primary': True, 'value': 'inactive@example.net'}],
                       'id': 'U0000'},
                      {'active': True,
                       'emails': [{'primary': True, 'value': 'allowed@example.net'}],
                       'id': 'U0001'}]

        fake_deactivate = {'active': False}
        fake_activate = {'active': True}

        mock_get_secret.return_value = 'nope'
        mock_slack_get_users.return_value = fake_users
        mock_slack_deactivate_user.return_value = fake_deactivate
        mock_slack_activate_user.return_value = fake_activate

        logger.info('Check if allowed user stays allowed')
        ret = driver.verify_slack_users({'allowed@example.net': 'test'})
        assert(ret is True)

        logger.info('Check if inactive user is re-activated')
        ret = driver.verify_slack_users({'inactive@example.net': 'test', 'allowed@example.net': 'test'})
        assert(ret is True)

        logger.info('Check if user without allowed group user is removed')
        ret = driver.verify_slack_users({})
        assert(ret is True)

        logger.info('Check if we fail correctly to activate users')
        mock_slack_activate_user.return_value = fake_deactivate
        ret = driver.verify_slack_users({'inactive@example.net': 'test'})
        assert(ret is False)

        logger.info('Check if we fail correctly to deactivate users')
        mock_slack_deactivate_user.return_value = fake_activate
        ret = driver.verify_slack_users({})
        assert(ret is False)


slack_driver/slack.py
#!/usr/bin/env python

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
# Copyright (c) 2018 Mozilla Corporation
# Contributors: Guillaume Destuynder <kang@mozilla.com>

# This file exists because: https://github.com/slackapi/python-slackclient/issues/292
# At this time there is no SCIM API library
# See also https://api.slack.com/scim#access

import http.client
import json


class SlackAPI(object):
    def __init__(self, token, uri="api.slack.com"):
        self.headers = {'content-type': "application/json",
                        'Authorization': "Bearer {}".format(token)}
        self.conn = http.client.HTTPSConnection(uri)

    def __del__(self):
        self.headers = None

    def _request(self, rtype, rpath, payload_json={}):
        """
        Wraps all API requests
        """
        self.conn.request(rtype, rpath, payload_json, self.headers)
        response = self.conn.getresponse()

        if (response.status >= 300) or (response.status < 200):
            raise Exception('HTTPCommunicationFailed', (response.status, response.reason, rpath, payload_json))

        ret = json.loads(response.read().decode('utf-8'))
        return ret

    def _depaginated_request(self, rtype, rpath, payload_json={}):
        """
        Returns all results from a paginated request
        """
        results = []

        res = self._depaginate(self._request(rtype, rpath, payload_json))
        results += res.get('results')
        while (not res.get('done')):
            res = self._depaginate(self._request(rtype, rpath, payload_json))
            results += res.get('results')
        return results

    def _depaginate(self, response):
        """
        Checks if we need to depaginate the request result, and return that fact in addition to the actual results of
        the request
        """
        # note that slack index starts at 1 instead of 0, so we fix that
        index = response.get('startIndex') - 1
        totals = response.get('totalResults')
        per_page = response.get('itemsPerPage')
        results = response.get('Resources')
        ret = {'done': False, 'results': results}

        # is there more pages left than total pages?
        # otherwise we're done
        if index + per_page >= totals:
            ret['done'] = True

        return ret

    def get_users(self, scim_filter=''):
        """
        Ex scim_filter: userName%20Eq%20kang_slack
        See also https://api.slack.com/scim#users
        and https://api.slack.com/scim#filter for the scim_filter syntax

        Returns list, similar to:
        [{'schemas': ['urn:scim:schemas:core:1.0'], 'id': 'UB0GWPDCM', 'externalId': '', 'meta': {'created':
        '2018-06-01T16:10:18-07:00', 'location': 'https://api.slack.com/scim/v1/Users/UB0GWPDCM'}, 'userName':
        'kang_slack', 'nickName': 'kang_slack', 'name': {'givenName': '', 'familyName': ''}, 'displayName': '',
        'profileUrl': 'https://mozilla-sandbox-scim.slack.com/team/kang_slack', 'title': '', 'timezone':
        'America/Los_Angeles', 'active': False, 'emails': [{'value': 'kang+slack@mozilla.com', 'primary': True}],
        'photos': [{'value':
        'https://example.net/test.png',
        'type': 'photo'}], 'groups': []}]

        """
        return self._depaginated_request("GET", "/scim/v1/Users?filter={}".format(scim_filter))

    def activate_user(self, slack_user_id):
        payload = {'active': True}
        payload_json = json.dumps(payload)
        return self._request("PATCH", "/scim/v1/Users/{}".format(slack_user_id), payload_json)

    def deactivate_user(self, slack_user_id):
        """
        See also https://api.slack.com/scim#users
        """
        payload = {'active': False}
        payload_json = json.dumps(payload)
        return self._request("PATCH", "/scim/v1/Users/{}".format(slack_user_id), payload_json)


slack_driver/vault.py
import boto3
import utils

from boto3.dynamodb.conditions import Attr

try:
    from settings import get_config
except ImportError:
    from slack_driver.settings import get_config


def setup_logging():
    global logger
    config = get_config()
    custom_logger = utils.CISLogger(
        name=__name__,
        level=config('logging_level', namespace='cis', default='INFO'),
        cis_logging_output=config('logging_output', namespace='cis', default='cloudwatch'),
        cis_cloudwatch_log_group=config('cloudwatch_log_group', namespace='cis', default='staging')
    ).logger()

    logger = custom_logger.get_logger()


class CISTable(object):
    def __init__(self, table_name):
        self.boto_session = boto3.session.Session()
        self.table_name = table_name
        self.table = None
        setup_logging()

    def connect(self):
        resource = self.boto_session.resource('dynamodb')
        self.table = resource.Table(self.table_name)
        return self.table

    @property
    def all(self):
        if self.table is None:
            self.connect()

        response = self.table.scan(FilterExpression=Attr('active').eq(True))

        users = response.get('Items')

        while 'LastEvaluatedKey' in response:
            response = self.table.scan(
                FilterExpression=Attr('active').eq(True),
                ExclusiveStartKey=response['LastEvaluatedKey']
            )
            users.extend(response['Items'])

        logger.debug('Returning total: {} users from the identity vault.'.format(len(users)))
        return users


class People(object):
    def __init__(self):
        self.config = get_config()
        self.table_name = self.config('dynamodb_person_table', namespace='cis', default='fake-identity-vault')
        self.table = CISTable(self.table_name)
        setup_logging()

    def people_in_group(self, groups):
        """Returns a dict of dicts for each user that matches a group from the list"""
        found_users = {}
        for user in self.table.all:
            for user_glist in user.get('groups', []):
                if user_glist in groups:
                    found_users[user.get('primaryEmail')] = user
                    break

        logger.debug('Returning total of : {} users for this run of the connector.'.format(
            len(found_users))
        )
        return found_users


slack_driver/driver.py
import credstash
import os
import utils
import requests
import slack
import yaml

from settings import get_config
from vault import People


def setup_logging(config):
    global logger
    custom_logger = utils.CISLogger(
        name=__name__,
        level=config('logging_level', namespace='cis', default='INFO'),
        cis_logging_output=config('logging_output', namespace='cis', default='cloudwatch'),
        cis_cloudwatch_log_group=config('cloudwatch_log_group', namespace='cis', default='staging')
    ).logger()

    logger = custom_logger.get_logger()


def get_access_rules(appsyml):
    """
    Fetch access rules
    @appsyml str URL
    returns dict
    """
    ## Sample appsyml return value format:
    ## { 'apps': [
    ##          {'application': {'name': 'Account Portal', 'op': 'auth0', 'url': 'https://login.mozilla.com/', 'logo':
    ##          'accountmanager.png', 'authorized_users': [], 'authorized_groups': ['team_moco', 'team_mofo'],
    ##          'display': True, ## 'vanity_url': ['/accountmanager']}}
    ##         ]
    ## }
    logger.debug('Fetching access rules.')
    r = requests.get(appsyml)
    if not r.ok:
        logger.warning('Failed to fetch access rules, will not deprovision users.')
        return []

    access_rules = yaml.load(r.text).get('apps')
    logger.debug('Received apps.yml size {}'.format(len(r.text)))
    return access_rules


def get_secret(secret_name, context):
    """Fetch secret from environment or credstash."""
    secret = os.getenv(secret_name.split('.')[1], None)

    if not secret:
        secret = credstash.getSecret(
            name=secret_name,
            context=context,
            region="us-west-2"
        )
    return secret


def verify_slack_users(allowed_users):
    """
    Find all Slack users which aren't allowed and deactivate them
    """
    # 1. list all slack users
    # 2. diff slack users with allowed_users
    # 3. disable all users in the diff and log

    slack_token = get_secret('slack-driver.token', {'app': 'slack-driver'}).split('\n')[0]
    sc = slack.SlackAPI(slack_token)

    slack_user_list = sc.get_users()
    logger.debug('Found {} Slack users in Slack database'.format(len(slack_user_list)))

    users = {}
    users_already_deactivated = {}
    for user in slack_user_list:
        # Find primary email from the list of user's email
        user_email = None
        for e in user.get('emails'):
            if e.get('primary'):
                user_email = e.get('value')
                logger.debug('Found primary email {}'.format(user_email))
                break

        # Just in case Slack API somehow doesn't have a primary email. This should normally not happen.
        if user_email is None:
            user_email = user.get('emails')[0].get('value')
            logger.warning('No slack primary email set for user {}, '
                           'selecting first email in list {}'.format(e.get('id'), user_email))

        if user.get('active'):
            users[user_email] = user.get('id')
        else:
            users_already_deactivated[user_email] = user.get('id')

    failure = 0
    users_to_enable = set(users_already_deactivated.keys()) & set(allowed_users.keys())
    logger.debug('Will re-enabled {} user(s) which were deactivated, '
                 'but got access to Slack again'.format(len(users_to_enable)))

    for u in users_to_enable:
        logger.debug('Will now enable {} ({})'.format(u, users_already_deactivated[u]))
        try:
            ret = sc.activate_user(users_already_deactivated[u])
            if (ret.get('active') is not True):
                logger.warning('Failed to enable user {}'.format(users_already_deactivated[u]))
                failure = failure + 1
        except Exception as e:
            logger.warning('Could not enable user {}, exception: {}'.format(users_already_deactivated[u], e))
            failure = failure + 1

    users_to_disable = set(users.keys()) - set(allowed_users.keys())
    logger.debug('Will disable {} user(s) which should no longer have access to Slack'.format(len(users_to_disable)))

    for u in users_to_disable:
        logger.debug('Will now disable user {} ({})'.format(u, users[u]))
        # sample return msg:
        # {"schemas": ["urn:scim:schemas:core:1.0"], "id": "UB0GWPDCM", "externalId": "", "meta": {"created":
        # "2018-06-01T16:10:18-07:00", "location": "https://api.slack.com/scim/v1/Users/UB0GWPDCM"}, "userName":
        # "kang_slack", "nickName": "kang_slack", "name": {"givenName": "", "familyName": ""}, "displayName": "",
        # "profileUrl": "https://mozilla-sandbox-scim.slack.com/team/kang_slack", "title": "", "timezone":
        # "America/Los_Angeles", "active": false, "emails": [{"value": "kang+slack@mozilla.com", "primary": true}],
        # "photos": [{"value":
        # "https://example.net/test.png",
        # "type": "photo"}], "groups": []}

        try:
            ret = sc.deactivate_user(users[u])
            if (ret.get('active') is not False):
                logger.warning('Failed to disable user {}'.format(users[u]))
                failure = failure + 1
        except Exception as e:
            # This can happen if the user is unmodifiable, for example if its a space owner
            logger.warning('Could not disable user {}, exception: {}'.format(users[u], e))
            failure = failure + 1

    if failure != 0:
        logger.warning('{} failure(s) occured'.format(failure))
        return False

    return True


def handle(event=None, context={}):
    config = get_config()
    setup_logging(config)
    logger.info('Initializing Slack driver.')

    appsyml = config('appsyml', namespace='slack_driver', default='https://cdn.sso.mozilla.com/apps.yml')
    slack_app = config('slack_app', namespace='slack_driver', default='Slack')

    access_rules = get_access_rules(appsyml)
    app = None
    authorized_groups = []
    for app in access_rules:
        actual_app = app.get('application')
        if actual_app.get('name') == slack_app:
            # XXX CIS with v1 profile prepend ldap groups with `ldap_` but the rest of the infra does not... so
            # workaround here:
            authorized_groups = []
            known_idp_prefix = ['mozilliansorg', 'hris', 'ldap']
            for g in actual_app.get('authorized_groups'):
                if g.split('_')[0] not in known_idp_prefix:
                    # its an ldap group
                    authorized_groups.append("ldap_" + g)
                else:
                    authorized_groups.append(g)

            logger.debug('Valid and authorized users are in groups {}'.format(authorized_groups))
            break

    if app is None:
        logger.warning('Did not find {} in access rules, will not deprovision users'.format(slack_app))
        return

    logger.debug('Searching DynamoDb for people.')
    people = People()

    logger.debug('Filtering person list to groups.')
    allowed_users = people.people_in_group(authorized_groups)
    logger.debug('Found {} Slack users which are allowed'.format(len(allowed_users)))

    logger.debug('Disable Slack users.')
    if not verify_slack_users(allowed_users):
        logger.warning('Failed to verify slack users - some users may not have been deprovisioned')


slack_driver/main.py
import credstash
import json
import os
import utils
import requests
import slack
import yaml

from settings import get_config
from vault import People

config = get_config()

custom_logger = utils.CISLogger(
    name=__name__,
    level=config('logging_level', namespace='cis', default='INFO'),
    cis_logging_output=config('logging_output', namespace='cis', default='cloudwatch'),
    cis_cloudwatch_log_group=config('cloudwatch_log_group', namespace='cis', default='staging')
).logger()

logger = custom_logger.get_logger()

def get_access_rules(appsyml):
    """
    Fetch access rules
    @appsyml str URL
    returns dict
    """
    ## Sample appsyml return value format:
    ## { 'apps': [
    ##            {'application': {'name': 'Account Portal', 'op': 'auth0', 'url': 'https://login.mozilla.com/', 'logo':
    ##            'accountmanager.png', 'authorized_users': [], 'authorized_groups': ['team_moco', 'team_mofo'],
    ##            'display': True, ## 'vanity_url': ['/accountmanager']}}
    ##           ]
    ## }
    logger.debug('Fetching access rules.')
    r = requests.get(appsyml)
    if not r.ok:
        logger.warning('Failed to fetch access rules, will not deprovision users.')
        return []

    access_rules = yaml.load(r.text).get('apps')
    logger.debug('Received apps.yml size {}'.format(len(r.text)))
    return access_rules

def get_secret(secret_name, context):
    """Fetch secret from environment or credstash."""
    secret = os.getenv(secret_name.split('.')[1], None)

    if not secret:
        secret = credstash.getSecret(
            name=secret_name,
            context=context,
            region="us-west-2"
        )
    return secret

def verify_slack_users(config, allowed_users):
    """
    Find all Slack users which aren't allowed and deactivate them
    """
    # 1. list all slack users
    # 2. diff slack users with allowed_users
    # 3. disable all users in the diff and log

    slack_token = get_secret('slack-driver.token', {'app': 'slack-driver'}).split('\n')[0]
    sc = slack.SlackAPI(slack_token)

    slack_user_list = sc.get_users()
    logger.debug('Found {} Slack users in Slack database'.format(len(slack_user_list)))

    users = {}
    for user in slack_user_list:
        # Find primary email from the list of user's email
        user_email = None
        for e in user.get('emails'):
            if e.get('primary'):
                user_email = e.get('value')
                logger.debug('Found primary email {}'.format(user_email))
                break

        # Just in case Slack API somehow doesn't have a primary email. This should normally not happen.
        if user_email is None:
            user_email = user.get('emails')[0].get('value')
            logger.warning('No slack primary email set for user {}, '
                           'selecting first email in list {}'.format(e.get('id'), user_email))

        users[user_email] = user.get('id')

    users_to_disable = set(users.keys()) - set(allowed_users.keys())
    logger.debug('Will disable {} user(s) which should no longer have access to Slack'.format(len(users_to_disable)))

    failure = 0
    for u in users_to_disable:
        logger.debug('Will now disable user {} ({})'.format(u, users[u]))
        # sample return msg:
        # {"schemas": ["urn:scim:schemas:core:1.0"], "id": "UB0GWPDCM", "externalId": "", "meta": {"created": "2018-06-01T16:10:18-07:00", "location": "https://api.slack.com/scim/v1/Users/UB0GWPDCM"}, "userName": "kang_slack", "nickName": "kang_slack", "name": {"givenName": "", "familyName": ""}, "displayName": "", "profileUrl": "https://mozilla-sandbox-scim.slack.com/team/kang_slack", "title": "", "timezone": "America/Los_Angeles", "active": false, "emails": [{"value": "kang+slack@mozilla.com", "primary": true}], "photos": [{"value": "https://secure.gravatar.com/avatar/8363a16c1147ee60fff6be4c8b30aaa1.jpg?s=192&d=https%3A%2F%2Fcfr.slack-edge.com%2F7fa9%2Fimg%2Favatars%2Fava_0009-192.png", "type": "photo"}], "groups": []}

        try:
            ret = sc.deactivate_user(users[u])
            if (ret.get('active') is not False):
                logger.warning('Failed to disable user {}'.format(users[u]))
                failure = failure + 1
        except Exception as e:
            # This can happen if the user is unmodifiable, for example if its a space owner
            logger.warning('Could not disable user {}, an exception occured: {}'.format(users[u], e))
            failure = failure + 1

    if failure != 0:
        logger.warning('{} failure(s) occured'.format(failure))
        return False

    return True

def handle(event=None, context={}):
    logger.info('Initializing Slack driver.')

    logger.debug('Getting configuration from environment.')
    config = get_config()

    filter_prefix = config('prefix', namespace='slack_driver', default='')
    driver_mode = config('interactive', namespace='slack_driver', default='True')
    environment = config('environment', namespace='slack_driver', default='development')
    appsyml = config('appsyml', namespace='slack_driver', default='https://cdn.sso.mozilla.com/apps.yml')
    slack_app = config('slack_app', namespace='slack_driver', default='Slack')


    access_rules = get_access_rules(appsyml)
    app = None
    authorized_groups = []
    for app in access_rules:
        actual_app = app.get('application')
        if actual_app.get('name') == slack_app:
            # XXX CIS with v1 profile prepend ldap groups with `ldap_` but the rest of the infra does not... so
            # workaround here:
            authorized_groups = []
            known_idp_prefix = ['mozilliansorg', 'hris', 'ldap']
            for g in actual_app.get('authorized_groups'):
                if g.split('_')[0] not in known_idp_prefix:
                    #its an ldap group
                    authorized_groups.append("ldap_"+g)
                else:
                    authorized_groups.append(g)

            logger.debug('Valid and authorized users are in groups {}'.format(authorized_groups))
            break

    if app == None:
        logger.warning('Did not find {} in access rules, will not deprovision users'.format(slack_app))
        return

    logger.debug('Searching DynamoDb for people.')
    people = People()

    logger.debug('Filtering person list to groups.')
    allowed_users = people.people_in_group(authorized_groups)
    logger.debug('Found {} Slack users which are allowed'.format(len(allowed_users)))

    logger.debug('Disable Slack users.')
    if not verify_slack_users(config, allowed_users):
        logger.warning('Failed to verify slack users - some users may not have been deprovisioned')
